static/app/api.tsx

```diff
@@ -581,3 +585,77 @@ export class Client {
     );
   }
 }
+
+const VALIDATE_SUBDOMAIN = /^[a-zA-Z0-9][a-zA-Z0-9-]*[^_-]$/;
+
+function generateOrganizationUrl(organization: OrganizationSummary): string | undefined {
+  const organizationBaseUrl = ConfigStore.get('organizationBaseUrl');
+  const organizationUrl = new URL('/', organizationBaseUrl);
+
+  const {slug} = organization;
+  if (!VALIDATE_SUBDOMAIN.test(slug)) {
+    return undefined;
+  }
+
+  organizationUrl.hostname = `${slug}.${organizationUrl.hostname}`.toLowerCase();
+
+  // Remove any and all trailing slashes
+  return organizationUrl.toString().replace(/\/*$/, '').toLowerCase();
+}
+
+function generateOrganizationBaseUrl(organizationUrl: string) {
+  return `${organizationUrl}${DEFAULT_BASE_URL}`;
+}
+
+type APIRouteType = 'organization-events';
+
+type APIRouterRendererArgs = {
+  organization: OrganizationSummary;
+  organizationUrl: string | undefined;
+  sentryUrl: string;
+  useOrganizationUrl: boolean;
+};
+type APIRouterRenderer = (args: APIRouterRendererArgs) => string;
+
+const routeRenderMap: Record<APIRouteType, APIRouterRenderer> = {
+  'organization-events': function ({organizationUrl, organization, useOrganizationUrl}) {
+    if (!organizationUrl || !useOrganizationUrl) {
+      return `/organizations/${organization.slug}/events/`;
+    }
+    return `${generateOrganizationBaseUrl(organizationUrl)}/events/`;
+  },
```

Comment:
Having to write a function for every API URL we want to generate is going to be challenging. Would we be able to split out the 'data' and 'logic'? By that I mean have something like:

```javascript
const routeMap = {
  'organization-events': [
    '/organizations/:slug/events/',
    '/events/',
  ],
};
```

We could then have a more general purpose function that formats data into the named route like you have with `resolveUrl()`